regparamrange crashes due to gsvd

The `regparamrange` employs the `deerlab.utils.gsvd` function which under certain circumstances leads to a crash. Until now crashes have been caused: 

- Numerical errors internally of Numpy:
````
~\AppData\Local\Programs\Python\Python37-32\lib\site-packages\numpy\core\fromnumeric.py in amax(a, axis, out, keepdims, initial, where)
   2704     """
   2705     return _wrapreduction(a, np.maximum, 'max', axis, None, out,
-> 2706                           keepdims=keepdims, initial=initial, where=where)
   2707 
   2708 

~\AppData\Local\Programs\Python\Python37-32\lib\site-packages\numpy\core\fromnumeric.py in _wrapreduction(obj, ufunc, method, axis, dtype, out, **kwargs)
     85                 return reduction(axis=axis, out=out, **passkwargs)
     86 
---> 87     return ufunc.reduce(obj, axis, dtype, out, **passkwargs)
     88 
     89 

ValueError: zero-size array to reduction operation maximum which has no identity
````

- Invalid values being sent to specific memory adresses in Scipy's Fortran subroutines:

````
/opt/hostedtoolcache/Python/3.8.6/x64/lib/python3.8/site-packages/scipy/linalg/decomp_qr.py:140: in qr
    qr, tau = safecall(geqrf, "geqrf", a1, lwork=lwork,
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

f = <fortran object>, name = 'geqrf'
args = (array([], shape=(100, 0), dtype=float64),)
kwargs = {'lwork': 0, 'overwrite_a': False}, lwork = None
ret = (array([], shape=(100, 0), dtype=float64), array([], dtype=float64), array([0.]), -7)

    def safecall(f, name, *args, **kwargs):
        """Call a LAPACK routine, determining lwork automatically and handling
        error return values"""
        lwork = kwargs.get("lwork", None)
        if lwork in (None, -1):
            kwargs['lwork'] = -1
            ret = f(*args, **kwargs)
            kwargs['lwork'] = ret[-2][0].real.astype(numpy.int_)
        ret = f(*args, **kwargs)
        if ret[-1] < 0:
>           raise ValueError("illegal value in %dth argument of internal %s"
                             % (-ret[-1], name))
E           ValueError: illegal value in 7th argument of internal geqrf
````
Thus far these errors seem to depend on the OS, the precision of the hardware and the BLAS linked to Numpy. 

Since neither `numpy` nor `scipy` have a working `gsvd` algorithm (that I could find), there are two options: 

- Invest time into debugging and fixing `deerlab.utils.gsvd`
- Deprecate `regparamrange`. The function has lost some utility now that the regularization parameter is optimized via Golden-search algorithms rather than grid searches. 

For now the best approach is to just patch `regparamrange` and catch the exception when running `gsvd(K,L)`. Once caught we can just run `svd(K)` as an approximating alternative.

Here is a MWE to reproduce the first error described above: 
```` python
import numpy as np 
import deerlab as dl
t = np.linspace(0,5,100)
r = np.linspace(2,6,150)
P = dl.dd_gauss(r,[4.5, 0.2])
K = dl.dipolarkernel(t,r)
V = K@P
fit = dl.fitsignal(V,t,r,'P',None,None,uqanalysis=False)
````
resulting in 
```` 
deerlab\fitsignal.py:425: in fitsignal
    fit = dl.fitregmodel(Vexp,Ks,r,regtype,regparam, weights=weights,uqanalysis=uqanalysis)
deerlab\fitregmodel.py:131: in fitregmodel
    alpha = dl.selregparam(V,K,r,regtype,regparam,regorder=regorder,weights=weights, nonnegativity=nonnegativity,huberparam=huberparam)
deerlab\selregparam.py:111: in selregparam
    alphaCandidates = dl.regparamrange(K,L)
deerlab\regparamrange.py:41: in regparamrange
    singularValues = gsvd(K,L)
deerlab\utils\utils.py:245: in gsvd
    U,_,_,C,S = csd(Q1,Q2)
deerlab\utils\utils.py:278: in csd
    V,U,Z,S,C = csd(Q2,Q1)
deerlab\utils\utils.py:311: in csd
    k = max(0, np.max(np.where(np.diag(C) <= 1/math.sqrt(2))))
<__array_function__ internals>:6: in amax
    ???
C:\Users\luis\AppData\Local\Programs\Python\Python36\lib\site-packages\numpy\core\fromnumeric.py:2706: in amax
    keepdims=keepdims, initial=initial, where=where)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _  

obj = (array([], dtype=int64),), ufunc = <ufunc 'maximum'>, method = 'max', axis = None, dtype = None, out = None, kwargs = {'initial': <no value>, 'keepdims': <no value>, 'where': <no value>}, passkwargs = {}

    def _wrapreduction(obj, ufunc, method, axis, dtype, out, **kwargs):
        passkwargs = {k: v for k, v in kwargs.items()
                      if v is not np._NoValue}

        if type(obj) is not mu.ndarray:
            try:
                reduction = getattr(obj, method)
            except AttributeError:
                pass
            else:
                # This branch is needed for reductions like any which don't
                # support a dtype.
                if dtype is not None:
                    return reduction(axis=axis, dtype=dtype, out=out, **passkwargs)
                else:
                    return reduction(axis=axis, out=out, **passkwargs)

>       return ufunc.reduce(obj, axis, dtype, out, **passkwargs)
E       ValueError: zero-size array to reduction operation maximum which has no identity

C:\Users\luis\AppData\Local\Programs\Python\Python36\lib\site-packages\numpy\core\fromnumeric.py:87: ValueError
````

--- a/deerlab/regparamrange.py
+++ b/deerlab/regparamrange.py
@@ -34,12 +34,8 @@ def regparamrange(K,L,noiselvl=0,logres=0.1):
     # Scaling by noise. This improves L curve corner detection for DEER.
     minmax_ratio = minmax_ratio*2**(noiselvl/0.0025)
 
-    try:
-        # Get generalized singular values of K and L
-        singularValues = gsvd(K,L)
-    except:
-        # Otherwise just estimate via the SVD from the dipolar kernel to avoid crash
-        _,singularValues,_ = svd(K)
+    # Get generalized singular values of K and L
+    singularValues = gsvd(K,L)
 
     DerivativeOrder = L.shape[1] - L.shape[0] # get order of derivative (=number of inf in singval)
     singularValues = singularValues[0:len(singularValues)-DerivativeOrder] # remove inf 

--- a/deerlab/utils/utils.py
+++ b/deerlab/utils/utils.py
@@ -249,7 +249,7 @@ def gsvd(A,B):
 #===============================================================================
 
 
-def csd(q1,q2):
+def csd(Q1,Q2):
 #===============================================================================
     """
     Cosine-Sine Decomposition
@@ -269,10 +269,10 @@ def csd(q1,q2):
     http://www.ar-tiste.com/m-fun/m-fun-index.html
 
     """
-    m,n = q1.shape
-    p,_ = q2.shape
+    m,n = Q1.shape
+    p,_ = Q2.shape
     if m < p:
-        s,c = csd(q2,q1)
+        s,c = csd(Q2,Q1)
         j = np.flip(np.arange(n)) 
         c = c[:,j] 
         s = s[:,j] 
@@ -284,7 +284,7 @@ def csd(q1,q2):
         s[np.arange(n),:] = s[i,:] 
         return c,s
 
-    _,sdiag,v = np.linalg.svd(q1)
+    _,sdiag,v = np.linalg.svd(Q1)
     c = np.zeros((m, n))
     np.fill_diagonal(c, sdiag)
     v = v.T.conj()
@@ -292,13 +292,13 @@ def csd(q1,q2):
     z = scp.linalg.hankel(z[:,n-1])
     c[0:n,:] = z@c[0:n,:]@z
     v = v@z
-    q2 = q2@v
+    Q2 = Q2@v
     k=0
     for j in range(1,n):
         if c[j,j] <= 1/np.sqrt(2): k=j
-    b = q2[:,0:k]
+    b = Q2[:,0:k]
     u2,r = np.linalg.qr(b,mode='complete')
-    s = u2.T@q2
+    s = u2.T@Q2
     t = np.minimum(p,n)
     tt = np.minimum(m,p)
     if k<t:

--- a/deerlab/utils/utils.py
+++ b/deerlab/utils/utils.py
@@ -232,148 +232,94 @@ def gsvd(A,B):
         QB = QB[:,0:p]
         n = p
 
-
     Q,_ = np.linalg.qr(np.vstack((A,B)), mode='reduced')
     Q1 = Q[0:m,0:p]
     Q2 = Q[m:m+n,0:p]
-    U,_,_,C,S = csd(Q1,Q2)
+    C,S = csd(Q1,Q2)
 
     # Vector of generalized singular values.
     q = min(m+n,p)
     # Supress divide by 0 warning        
     with warnings.catch_warnings():
         warnings.simplefilter('ignore')
-        U = np.vstack((np.zeros((q-m,1),'double'), np.diag(C,max(0,q-m)).reshape(len(np.diag(C,max(0,q-m))),1))) /  np.vstack((np.diag(S,0).reshape(len(np.diag(S,0)),1), np.zeros((q-n,1),'double') ))
+        U = np.vstack((np.zeros((q-m,1),'double'), np.diag(C,max(0,q-m)).reshape(len(np.diag(C,max(0,q-m))),1)))/np.vstack((np.diag(S,0).reshape(len(np.diag(S,0)),1), np.zeros((q-n,1),'double') ))
 
 
     return U
 #===============================================================================
 
 
-def csd(Q1,Q2):
+def csd(q1,q2):
 #===============================================================================
     """
     Cosine-Sine Decomposition
-
-    U,V,Z,C,S = csd(Q1,Q2)
-
-    Given Q1 and Q2 such that Q1'@Q1 + Q2'@Q2 = I, the
-    C-S Decomposition is a joint factorization of the form
-    Q1 = U@C@Z' and Q2=V@S@Z'
-    where U, V, and Z are orthogonal matrices and C and S
-    are diagonal matrices (not necessarily square) satisfying
-    C'@C + S'@S = I
-    """
-    [m,p] = np.shape(Q1)
-    n = np.shape(Q2)[0]
-
-
-    if m < n:
-        V,U,Z,S,C = csd(Q2,Q1)
-        j = np.flip(np.arange(p)) 
-        C = C[:,j] 
-        S = S[:,j] 
-        Z = Z[:,j]
-        m = min(m,p) 
-        i =  np.flip(np.arange(m)) 
-        C[np.arange(m),:] = C[i,:] 
-        U[:,np.arange(m)] = U[:,i]
-        n = min(n,p) 
-        i =  np.flip(np.arange(n)) 
-        S[np.arange(n),:] = S[i,:] 
-        V[:,np.arange(n)] = V[:,i]
-        return U,V,Z,C,S
+    -------------------------
     
-    U,sdiag,VH = np.linalg.svd(Q1)
-    C= np.zeros((m, p))
-    np.fill_diagonal(C, sdiag)
-    Z = VH.T.conj()
+    Given Q1 and Q2 such that Q1'* Q1 + Q2'* Q2 = I, the
+    C-S Decomposition is a joint factorization of the form
+          Q1 = U1*C*V' and Q2=U2*S*V'
+    where U1,U2,V are orthogonal matrices and C and S are diagonal
+    matrices (not necessarily square) satisfying
+             C'* C + S'* S = I
+    The diagonal entries of C and S are nonnegative and the
+    diagonal elements of C are in nondecreasing order.
+    The matrix Q1 cannot have more columns than rows.
 
-    q = min(m,p)
-    i = np.arange(0, q, 1)
-    j = np.arange(q-1, -1, -1)
-    C[i,i] = C[j,j]
-    U[:,i] = U[:,j]
-    Z[:,i] = Z[:,j]
-    S = Q2@Z
+    Based on the Octave code by Artiste (submitted by S.J.Leon): 
+    http://www.ar-tiste.com/m-fun/m-fun-index.html
 
-    if q == 1:
-        k = 0
-    elif m < p:
-        k = n
-    else:
-        k = max(0, np.max(np.where(np.diag(C) <= 1/math.sqrt(2))))
-    V,_ = scp.linalg.qr(S[:,0:k])
-
-    S = V.T@S
-    r = min(k,m)
-    S[:,0:r-1] = diagf(S[:,0:r-1])
-
-    if (m == 1 and p > 1):
-        S[0,0] = 0
-
-    if k < min(n,p):
-        r = min(n,p)
-        i = np.arange(k, n, 1)
-        j = np.arange(k, r, 1)
-        [UT,STdiag,VH] = np.linalg.svd(S[np.ix_(i,j)])
-        ST = np.zeros((len(i), len(j)))
-        np.fill_diagonal(ST, STdiag)
-        VT = VH.T.conj()
-        if k > 0:
-            S[0:k,j] = 0
-        S[np.ix_(i,j)] = ST
-        C[:,j] = C[:,j]@VT
-        V[:,i] = V[:,i]@UT
-        Z[:,j] = Z[:,j]@VT
-        i = np.arange(k, q, 1)
-        [Q,R] = scp.linalg.qr(C[np.ix_(i,j)])
-        C[np.ix_(i,j)] = diagf(R)
-        U[:,i] = U[:,i]@Q
-
+    """
+    m,n = q1.shape
+    p,_ = q2.shape
     if m < p:
-        # Diagonalize final block of S and permute blocks.
-        eps = np.finfo(float).eps
-        q = min([np.count_nonzero(abs(np.diag(C,0))>10*m*eps),
-                np.count_nonzero(abs(np.diag(S,0))>10*n*eps),
-                np.count_nonzero(np.amax(abs(S[:,m+1:p]),axis = 1)<np.sqrt(eps))])
+        s,c = csd(q2,q1)
+        j = np.flip(np.arange(n)) 
+        c = c[:,j] 
+        s = s[:,j] 
+        m = np.minimum(m,p) 
+        i =  np.flip(np.arange(m)) 
+        c[np.arange(m),:] = c[i,:] 
+        n = np.minimum(n,p) 
+        i =  np.flip(np.arange(n)) 
+        s[np.arange(n),:] = s[i,:] 
+        return c,s
 
-        # maxq: maximum size of q such that the expression used later on,
-        #        i = [q+1:q+p-m, 1:q, q+p-m+1:n],
-        # is still a valid permutation.
-        maxq = m + n - p
-        q = q + np.count_nonzero(np.amax(abs(S[:,q+1:maxq+1]),axis = 1)>np.sqrt(eps))
+    _,sdiag,v = np.linalg.svd(q1)
+    c = np.zeros((m, n))
+    np.fill_diagonal(c, sdiag)
+    v = v.T.conj()
+    z = np.eye(n,n)
+    z = scp.linalg.hankel(z[:,n-1])
+    c[0:n,:] = z@c[0:n,:]@z
+    v = v@z
+    q2 = q2@v
+    k=0
+    for j in range(1,n):
+        if c[j,j] <= 1/np.sqrt(2): k=j
+    b = q2[:,0:k]
+    u2,r = np.linalg.qr(b,mode='complete')
+    s = u2.T@q2
+    t = np.minimum(p,n)
+    tt = np.minimum(m,p)
+    if k<t:
+        r2 = s[np.ix_(range(k,p),range(k,t))]
+        _,sdiag,vt = np.linalg.svd(r2)
+        ss= np.zeros(r2.shape)
+        np.fill_diagonal(ss, sdiag)
+        vt = vt.T.conj()
+        s[k:p,k:t] = ss
+        c[:,k:t] = c[:,k:t]@vt
+        w = c[k:tt,k:t]
+        z,r = np.linalg.qr(w,mode='complete')
+        c[k:tt,k:t] = r
+    for j in range(n):
+        if c[j,j]<0:
+            c[j,j]  = -c[j,j]
+    for j in range(t):
+        if s[j,j]<0:
+            s[j,j]  = -s[j,j]
 
-        i = np.arange(q,n,1)
-        j = np.arange(m,p,1)
-        # At this point, S(i,j) should have orthogonal columns and the
-        # elements of S(:,q+1:p) outside of S(i,j) should be negligible.
-        Q,R = scp.linalg.qr(S[np.ix_(i,j)])
-        S[:,q+1:p] = 0
-        S[np.ix_(i,j)] = diagf(R)
-        V[:,i] = V[:,i]@Q
-        if n > 1:
-            i = np.concatenate((np.arange(q,q+p-m,1), np.arange(0,q,1), np.arange(q+p-m,n,1)))
-        else:
-            i = 1
-        j = np.concatenate((np.arange(m,p,1), np.arange(0,m,1)))
-        C = C[:,j]
-        S = S[np.ix_(i,j)]
-        Z = Z[:,j]
-        V = V[:,i]
-
-    if n < p:
-        #Final block of S is negligible.
-        S[:,n+1:p] = 0
-
-    # Make sure C and S are real and positive.
-    U,C = diagp(U,C,max(0,p-m))
-    C = C.real
-
-    V,S = diagp(V,S,0)
-    S = S.real
-
-    return U,V,Z,C,S
+    return c,s
 #===============================================================================
 
 #===============================================================================

